server/lib/db/index.js

```diff
@@ -289,6 +289,50 @@ class DatabaseConnector {
       .catch(next)
       .done();
   }
+
+  async paginateQuery(sql, params, tables, filters) {
+    let pager = {};
+    let rows = [];
+    let fetchAllData = false;
+
+    if (!params.limit) {
+      params.limit = 100;
+    } else if (params.limit && parseInt(params.limit, 10) === -1) {
+      fetchAllData = true;
+      delete params.limit;
+    }
+
+    if (params.page && parseInt(params.page, 10) === 0) {
+      delete params.page;
+    }
+
+    const queryParameters = filters.parameters();
+
+    if (fetchAllData) {
+      // fetch all data
+      const query = filters.applyQuery(sql.concat(' ', tables));
+      rows = await this.exec(query, queryParameters);
+    } else {
+      // paginated data
+
+      // FIXME: Performance issue, use SQL COUNT in a better way
+      const total = (await this.exec(filters.getAllResultQuery(sql.concat(' ', tables)), queryParameters)).length;
+      const page = params.page ? parseInt(params.page, 10) : 1;
+      const limit = params.limit ? parseInt(params.limit, 10) : 100;
```

Comment:
I suggest computing `page_count` here.  And when computing `page_min` and `page_max` below, if `page` > `page_count`, set them to `null' since they are invalid for pages beyond the max.  And apply similar fixes to the FilterParser below.

Reply:
We have to set `page_min` and `page_max` after the execution of the query since the `page_min` is sent to the SQL query in the `OFFSET` part.